File: ChartLib/src/main/java/com/yxc/chartlib/view/BaseChartRecyclerView.java
package com.yxc.chartlib.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.xiaomi.fitness.chart.attrs.BaseChartAttrs;
import com.xiaomi.fitness.chart.barchart.itemdecoration.BaseChartItemDecoration;
import com.xiaomi.fitness.chart.component.BaseYAxis;
import com.xiaomi.fitness.chart.component.VarietyMaxYAxis;
import com.xiaomi.fitness.chart.component.YAxis;
import com.xiaomi.fitness.common.utils.AppUtil;
import com.xiaomi.fitness.common.utils.DisplayUtil;


/**
 * @author yxc
 * @since  2019-05-10
 *
 */
public abstract class BaseChartRecyclerView<T extends BaseChartAttrs, I extends BaseChartItemDecoration> extends RecyclerView {

    public OnChartTouchListener onChartTouchListener;
    public T mAttrs;
    public I mItemDecoration;

    @Override
    public void addItemDecoration(@NonNull ItemDecoration decor) {
        super.addItemDecoration(decor);
        if (decor instanceof  BaseChartItemDecoration){
            mItemDecoration = (I) decor;
        }
    }

    public <Y extends BaseYAxis>void setYAxis(Y mYAxis){
        mItemDecoration.setYAxis(mYAxis);
    }

    public I getItemDecoration(){
        return mItemDecoration;
    }

    public BaseChartRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        this.mAttrs = getAttrs(context, attrs);
//        setRecyclerViewDefaultPadding();
    }

    @Override
    public boolean onTouchEvent(MotionEvent e) {
        if (e.getActionMasked() == MotionEvent.ACTION_DOWN) {
            if (onChartTouchListener != null){
                onChartTouchListener.onChartGestureStart(e);
            }
        } else if (e.getAction() == MotionEvent.ACTION_UP
                || e.getAction() == MotionEvent.ACTION_CANCEL) {
            if (onChartTouchListener != null) {
                onChartTouchListener.onChartGestureEnd(e);
            }
        } else if (e.getActionMasked() == MotionEvent.ACTION_MOVE){
            if (onChartTouchListener != null) {
                onChartTouchListener.onChartGestureMovingOn(e);
            }
        }
        return super.onTouchEvent(e);
    }

    public void setOnChartTouchListener(OnChartTouchListener onChartTouchListener) {
        this.onChartTouchListener = onChartTouchListener;
    }

    private void setRecyclerViewDefaultPadding() {
        int paddingLeft = getPaddingLeft();
        int paddingRight = getPaddingRight();
        if (mAttrs.enableStartYAxisLabel) {
            paddingRight = DisplayUtil.dip2px(36);
        }
        if (mAttrs.enableEndYAxisLabel) {
            paddingLeft = DisplayUtil.dip2px(36);
        }
        int paddingStart = AppUtil.isRTLDirection() ? paddingRight : paddingLeft;
        int paddingEnd = AppUtil.isRTLDirection() ? paddingLeft : paddingStart;
        setPadding(paddingStart, getPaddingTop(), paddingEnd, getPaddingBottom());
    }


    @Override
    public boolean fling(int velocityX, int velocityY) {
        velocityX *= mAttrs.ratioVelocity;
        velocityY *= mAttrs.ratioVelocity;
        return super.fling(velocityX, velocityY);
    }



    public interface OnChartTouchListener {

        void onChartGestureStart(MotionEvent e);

        void onChartGestureEnd(MotionEvent e);

        void onChartGestureMovingOn(MotionEvent e);
    }

    protected abstract T getAttrs(@NonNull Context context, @Nullable AttributeSet attrs);

}

File: ChartLib/src/main/java/com/yxc/chartlib/view/BaseChartRecyclerView.kt
package com.yxc.chartlib.view

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.recyclerview.widget.RecyclerView
import com.yxc.chartlib.attrs.BaseChartAttrs
import com.yxc.chartlib.barchart.itemdecoration.BaseChartItemDecoration

/**
 * @author yxc
 * @since  2019-05-10
 */
abstract class BaseChartRecyclerView<T : BaseChartAttrs, I : BaseChartItemDecoration<*, *>>(
    context: Context,
    attrs: AttributeSet
) : RecyclerView(context, attrs) {

    private lateinit var onChartTouchListener: OnChartTouchListener
    @JvmField
    var mAttrs: T
    var mItemDecoration: I? = null
    override fun addItemDecoration(decor: ItemDecoration) {
        super.addItemDecoration(decor)
        if (decor is BaseChartItemDecoration<*, *>) {
            mItemDecoration = decor as I
        }
    }

    init {
        mAttrs = getAttrs(context, attrs)
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        if (e.actionMasked == MotionEvent.ACTION_DOWN) {
            onChartTouchListener.onChartGestureStart(e)
        } else if (e.action == MotionEvent.ACTION_UP
            || e.action == MotionEvent.ACTION_CANCEL
        ) {
            onChartTouchListener.onChartGestureEnd(e)
        } else if (e.actionMasked == MotionEvent.ACTION_MOVE) {
            onChartTouchListener.onChartGestureMovingOn(e)
        }
        return super.onTouchEvent(e)
    }

    override fun fling(velocityX: Int, velocityY: Int): Boolean {
        val velocityXF = velocityX * mAttrs.ratioVelocity
        val velocityYF = velocityY * mAttrs.ratioVelocity
        return super.fling(velocityXF.toInt(), velocityYF.toInt())
    }

    fun addOnChartTouchListener(onChartTouchListener: OnChartTouchListener) {
        this.onChartTouchListener = onChartTouchListener
    }

    interface OnChartTouchListener {
        fun onChartGestureStart(e: MotionEvent)
        fun onChartGestureEnd(e: MotionEvent)
        fun onChartGestureMovingOn(e: MotionEvent)
    }

    protected abstract fun getAttrs(context: Context, attrs: AttributeSet): T
}